### DISCUSSION TITLE
Orientation matrix and Abaqus inp file

### DISCUSSION THREAD

**Discussion by xinliu09** [neperfepx/neper#130] [Q&A] [answered]
Hi,

First of all, thanks for developing such a great tool.
I am trying to generate a Abaqus inp file from neper with a ramdon orientation. I have two question:
(1) I use this command "neper -T -n 10 -ori random -oridescriptor rotmat -format tess,ori" to generate the orientation file. If the 9 numbers in each line in ori file denotes x1,x2,x3,...,x9. Will the rotation matrix be R={{x1,x2,x3},{x4,x5,x6},{x7,x8,x9}}? and 
(2) Is that possible to automatically define the orientation information in inp file? Currently, I can use "neper -M filename.tess -format inp" to generate inp file, but there is no orientation information in the inp file. 

Thank you,
Xin Liu

**Comment by rquey** [comment, marked as answer]
(1) Correct.
(2) Afaik, Abaqus requires a UMAT to do crystal plasticity and would not support the definition of grain orientations in the .inp file.  This is why you have to do it separately (output the orientations using `-T -for ori` and use this to write your Abaqus file).

**Reply by xinliu09** [reply]
Got it. Thank you for the prompt responses.

**Comment by sujit56438** [comment]
I am trying to find ‘Euler angle’ of each grain for finding rotation matrix of each grain where the laser light passes through the crystal.
sir, can u guide me how can i get it

**Reply by rquey** [reply]
Use `-statcell euler-bunge` (to get it in a "statistics" file) or `-for tess,ori -orides euler-bunge` (to get it in an "orientation" file).

**Reply by sujit56438** [reply]
@rquey  sir,
How can I get the exact values of the grain diameters on the path along the laser beam that are extracted from the modeling data, as shown in figure bellow:

![Screenshot (35)](https://github.com/neperfepx/neper/assets/142901058/d9dc787d-1919-4efc-86fe-ed697377b70b)

Currently, I am using for 500x500x500 micrometer crystal
 neper -T -n 350 -morpho "size:lognormal(mean=95,sigma=48),gg" -domain "cube(500,500,500)"  -statcell euler-bunge 
![image1](https://github.com/neperfepx/neper/assets/142901058/45f12b92-4ba1-45cb-aaae-4724140a0d7d)

**Reply by rquey** [reply]
Create your path as a discrete point file, and extract the cell to which each point belongs. Here is an example:
```
neper -T -n 10
seq 0.005 0.01 1 | awk '{print $1,0.5,0.5}' > path
neper -T -loadtess n10-id1.tess -loadpoint path -statpoint cell

========================    N   e   p   e   r    =======================
Info   : A software package for polycrystal generation and meshing.
Info   : Version 4.7.0
Info   : Built with: gsl|muparser|opengjk|openmp|nlopt|libscotch (full)
Info   : Running on 20 threads.
Info   : <https://neper.info>
Info   : Copyright (C) 2003-2022, and GNU GPL'd, by Romain Quey.
Info   : Loading initialization file `/home/rquey/.neperrc'...
Info   : ---------------------------------------------------------------
Info   : MODULE  -T loaded with arguments:
Info   : [ini file] -tesrformat ascii
Info   : [com line] -loadtess n10-id1.tess -loadpoint path -statpoint
         cell
Info   : ---------------------------------------------------------------
Info   : Reading input data...
Info   : Importing tessellation...
Info   :     [i] Parsing file `n10-id1.tess'...
Info   :     [i] Parsed file `n10-id1.tess'.
Info   : Importing points...
Info   :     [i] Parsing file `path'...
Info   :     [i] Parsed file `path'.
Info   : Writing results...
Info   :     [o] Writing file `n10-id1.tess'...
Info   :     [o] Wrote file `n10-id1.tess'.
Info   : Writing statistics...
Info   :     [o] Writing file `n10-id1.stpoint'...
Info   :     [o] Wrote file `n10-id1.stpoint'.
Info   : Elapsed time: 0.005 secs.
========================================================================

```

`n10-id1.stpoint`:
```
7
7
7
7
7
7
7
7
7
7
7
5
5
5
5
5
5
5
5
5
5
5
5
10
10
10
10
10
10
10
10
10
10
10
10
10
10
10
10
10
10
10
10
10
10
10
10
10
10
10
10
10
10
10
10
10
10
10
10
10
10
10
10
10
10
10
10
10
10
10
10
10
10
10
10
10
10
10
10
10
10
10
10
10
10
10
10
10
9
9
9
9
9
9
9
9
9
9
9
9
```
